OutOfSchool/OutOfSchool.AuthCommon/Services/ProviderAdminService.cs

```diff
@@ -277,14 +306,56 @@ async Task<ResponseDto> UpdateProviderAdminOperation(UpdateProviderAdminDto upda
                     return response;
                 }
 
+                var oldWorkshopsIds = providerAdmin.ManagedWorkshops.Select(x => x.Id).ToList();
+
                 providerAdmin.ManagedWorkshops = !providerAdmin.IsDeputy
                     ? context.Workshops.Where(w => updateDto.ManagedWorkshopIds.Contains(w.Id)).ToList()
                     : new List<Workshop>();
 
                 await providerAdminRepository.Update(providerAdmin).ConfigureAwait(false);
+                var newWorkshopsIds = providerAdmin.ManagedWorkshops.Select(x => x.Id).ToList();
+                var addedWorkshopsIds = newWorkshopsIds.Except(oldWorkshopsIds).ToList();
+                var removedWorkshopsIds = oldWorkshopsIds.Except(newWorkshopsIds).ToList();
 
-                await providerAdminChangesLogService.SaveChangesLogAsync(providerAdmin, userId, OperationType.Update)
-                    .ConfigureAwait(false);
+                foreach (var addedId in addedWorkshopsIds)
+                {
+                    await providerAdminChangesLogService.SaveChangesLogAsync(
+                        providerAdmin,
+                        userId,
+                        OperationType.Update,
+                        "WorkshopId",
+                        string.Empty,
+                        addedId.ToString())
+                        .ConfigureAwait(false);
+                }
+
+                foreach (var removedId in removedWorkshopsIds)
+                {
+                    await providerAdminChangesLogService.SaveChangesLogAsync(
+                        providerAdmin,
+                        userId,
+                        OperationType.Update,
```

Comment:
(Logic) Should it be `OperationType.Update` ?
As for now I can't distinguish if some workshopId was added or removed.
Probably it could be coded not via `OperationType` but by id itself? (I.e. adding minus sign before removed ids - `"-" + removedId.ToString()`?)

Reply:
My logic is next. We have properties OperationType, PropertyName, OldValue, NewValue. 
If we update ProviderAdmin with by adding only one workshop - OperationType is "Update", PropertyName is "WorkshopId", OldValue is "string.Empy", NewValue is GUID of added workshop.
If we update ProviderAdmin with by removing only one workshop - OperationType is "Update", PropertyName is "WorkshopId", but OldValue is GUID of removed workshop, NewValue is string.Empty.

Reply:
I see, thank you. Then there is other question. 
Few lines below you are logging changes from `newPropertiesValues`. 
Let's imagine that `user.MiddleName` was changed from some value to empty/null string. Am I right that such change will never come to the log?

Reply:
You're right, it's because of the null check in the GetTrackedUserProperties method for propertyValue - changes with null value never come to log.